common/determined_common/schemas/schema.py

```diff
@@ -0,0 +1,431 @@
+    """
+
+    def __new__(cls: type, name: str, bases: Tuple, dct: dict) -> Any:
+        # Nothing to do on either the SchemaBase class or if there is no __init__ function.
+        if name == "SchemaBase" or "__init__" not in dct:
+            return super().__new__(cls, name, bases, dct)  # type: ignore
+
+        dct["__init__"] = auto_init(dct["__init__"])
+
+        return super().__new__(cls, name, bases, dct)  # type: ignore
+
+
+T = TypeVar("T", bound="SchemaBase")
+
+
+class SchemaBase(metaclass=AutoInit):
+    _id: str
+
+    def __init__(self, **kwargs: dict) -> None:
+        raise NotImplementedError
+
+    @classmethod
+    def from_none(cls: Type[T]) -> Optional[T]:
+        """
+        from_none is called inside from_dict, when a key is present as a literal None.
+
+        For most objects (ResourcesConfig, for example), a None value means it is not present.
+
+        However, some values (Hyperparameter, for example), a None value represents a real object.
+        This classmethod makes it possible to customize behavior in those situations.
+        """
+        return None
+
+    @classmethod
+    def from_dict(cls: Type[T], d: dict, prevalidated: bool = False) -> T:
+        if not isinstance(d, dict) or any(not isinstance(k, str) for k in d):
+            raise ValueError("from_dict() requires an input dictionary with only string keys")
+
+        # Validate before parsing.
+        if not prevalidated:
+            errors = expconf.validation_errors(d, cls._id)
+            if errors:
+                raise TypeError("\n".join(errors))
+
+        additional_properties_anno = cls.__annotations__.get("_additional_properties")
+
+        init_args = {}
+
+        # For every key in the dictionary, get the type from the class annotations.  If it is a
+        # sublcass of SchemaBase, call from_dict() or from_none() on it based on the value in the
+        # input.  Otherwise, make sure a primitive type and pass the value to __init__ directly.
+        for name, value in d.items():
+            anno = cls.__annotations__.get(name, additional_properties_anno)
+            if anno is None:
+                raise TypeError(
+                    f"from_dict() found a key '{name}' input which has no annotation.  This is a "
+                    "bug; all SchemaBase subclasses must have annotations which match the json "
+                    "schema definitions which they correspond to."
+                )
+            # Create an instance based on the type annotation.
+            init_args[name] = instance_from_annotation(anno, value, prevalidated=True)
+
+        return cls(**init_args)
+
+    def to_dict(self) -> dict:
+        return {k: to_dict(v) for k, v in vars(self).items()}
+
+    def fill_defaults(self) -> None:
+        # Create any non-present child objects.
+        for name, anno in self.__annotations__.items():
+            # Ignore special annotations.
+            if name.startswith("_"):
```

Comment:
Correct.  That wouldn't make sense, because by default there *aren't* additional properties.

The additional properties themselves can have defaults though, but that's not related to this.

Separately, in the follow-up work, I ended up removing `_additional_properties` concept entirely, and now just use dictionaries instead of `Schema` objects with `_additional_properties`.  That's not worth backporting to this PR, IMO.